runtime/intern.c

```diff
@@ -239,446 +284,395 @@ static void readfloats(struct caml_intern_state* s,
+                { caml_handle_gc_interrupt_no_async_exceptions(); } );
+    s->intern_dest = (header_t *) Hp_val(v);
+  } else {
+    CAMLassert (s->intern_dest == NULL);
   }
-  STACK_VAL(s->curr_vals, s->sp) = v;
-  STACK_FIELD(s->curr_vals, s->sp) = Val_int(field);
-  STACK_OP(s->curr_vals, s->sp) = Val_int(op);
-  STACK_ARG(s->curr_vals, s->sp) = Val_long(arg);
-  s->sp++;
-}
-
-static void stack_pop(struct intern_stack* s) {
-  CAMLassert(!stack_is_empty(s));
-  s->sp--;
-}
-
-static value stack_curr_val(struct intern_stack* s) {
-  CAMLassert(!stack_is_empty(s));
-  return STACK_VAL(s->curr_vals, s->sp - 1);
-}
-
-static int stack_curr_op(struct intern_stack* s) {
-  CAMLassert(!stack_is_empty(s));
-  return Int_val(STACK_OP(s->curr_vals, s->sp - 1));
-}
-
-static int stack_curr_field(struct intern_stack* s) {
-  CAMLassert(!stack_is_empty(s));
-  return Int_val(STACK_FIELD(s->curr_vals, s->sp - 1));
-}
-
-static intnat stack_curr_arg(struct intern_stack* s) {
-  CAMLassert(!stack_is_empty(s));
-  return Int_val(STACK_ARG(s->curr_vals, s->sp - 1));
-}
-
-static void stack_advance_field(struct intern_stack* s) {
-  intnat field, length;
-
-  CAMLassert(!stack_is_empty(s));
-
-  length = Int_val(STACK_ARG(s->curr_vals, s->sp - 1));
-  length--;
-  if (length == 0) {
-    stack_pop(s);
+  s->obj_counter = 0;
+  if (num_objects > 0) {
+    s->intern_obj_table =
+      (value *) caml_stat_alloc_noexc(num_objects * sizeof(value));
+    if (s->intern_obj_table == NULL) {
+      intern_cleanup(s);
+      caml_raise_out_of_memory();
+    }
   } else {
-    STACK_ARG(s->curr_vals, s->sp - 1) = Val_int(length);
-    field = Int_val(STACK_FIELD(s->curr_vals, s->sp - 1));
-    field++;
-    STACK_FIELD(s->curr_vals, s->sp - 1) = Val_int(field);
+    CAMLassert(s->intern_obj_table == NULL);
   }
-}
 
-static void stack_push_items(struct intern_stack* s, value dest, int n) {
-  if (n > 0) {
-    stack_push(s, dest, 0, OReadItems, n);
-  }
+  return;
 }
 
-static void intern_cleanup(struct caml_intern_state* s)
+static value intern_alloc_obj(struct caml_intern_state* s, caml_domain_state* d,
+                              mlsize_t wosize, tag_t tag)
 {
-  s->intern_src = NULL;
-  if (s->intern_input != NULL) {
-    caml_stat_free (s->intern_input);
-    s->intern_input = NULL;
-  }
-
-  if (s->stack)
-    stack_free(s->stack);
-  s->stack = NULL;
-
-  if (s->prev != NULL) {
-    CAMLassert (Caml_state->intern_state == s);
-    Caml_state->intern_state = s->prev;
+  void* p;
+
+  if (s->intern_dest) {
+    CAMLassert ((value*)s->intern_dest >= d->young_start &&
+                (value*)s->intern_dest < d->young_end);
+    p = s->intern_dest;
+    *s->intern_dest = Make_header (wosize, tag, 0);
+    s->intern_dest += 1 + wosize;
+  } else {
+    p = caml_shared_try_alloc(d->shared_heap, wosize, tag, 0 /* not pinned */);
+    d->allocated_words += Whsize_wosize(wosize);
+    if (p == NULL) {
+      caml_raise_out_of_memory();
```

Comment:
We should call `intern_cleanup(s)` before raising, right?

Reply:
Yes. That's correct.